vllm/compilation/fusion.py

```diff
@@ -0,0 +1,167 @@
+    register_replacement(rms_pattern_static, rms_replacement_static, inputs, fwd_only, my_patterns)
+
+    matches = []
+
+    def record_match_fn(match: Match):
+        matches.append(match)
+        return False
+
+    # with residual
+    inputs = [empty_fp8(5, 4), empty_bf16(5, 4), empty_bf16(5, 4), empty_bf16(1, 5), torch.empty(1, 1, device="cuda", dtype=torch.float32)]
+    register_replacement(rms_pattern_residual_static, rms_replacement_residual_static, inputs, fwd_only, my_patterns,
+                         extra_check=record_match_fn)
+
+    return my_patterns, matches
+
+
+def process_matches(matches, graph: torch.fx.Graph):
+    for match in matches:
+        nodes = list(graph.nodes)
+        # TODO this is an expensive check
+        if not all(node in nodes for node in match.nodes):
+            raise ValueError(
+                f"Broken match: not all nodes in graph: {[node for node in match.nodes if node not in nodes]}")
+        last_node_in_match = max(match.nodes, key=lambda x: nodes.index(x))
+        with graph.inserting_after(last_node_in_match):
+            kwargs = match.kwargs
+            kwargs["epsilon"] = 1e-5
+
+            fused_node = graph.call_function(auto_functionalized,
+                                             (torch.ops._C.fused_add_rms_norm_static_fp8_quant.default,),
+                                             kwargs=kwargs)
+
+            graph.inserting_after(fused_node)
+            result_node_new = graph.call_function(operator.getitem, (fused_node, 1))
+            residual_node_new = graph.call_function(operator.getitem, (fused_node, 2))
+
+        # find the output and the residual
+        def find_auto_fn(op):
+            for node in match.nodes:
+                if node.op == "call_function" and node.target == auto_functionalized and node.args[0] == op:
+                    return node
+            return None
+
+        def find_getitem(node, idx):
+            for user in node.users:
+                if user.op == "call_function" and user.target == operator.getitem and user.args[1] == idx:
+                    return user
+            return None
+
+        rms_node = find_auto_fn(torch.ops._C.fused_add_rms_norm.default)
+        quant_node = find_auto_fn(torch.ops._C.static_scaled_fp8_quant.default)
+        assert rms_node is not None
+        assert quant_node is not None
+
+        assert len(rms_node.users) == 2
+        assert len(quant_node.users) == 1
+
+        # meta["val"] is used by de-functionalization
+        rms_val = rms_node.meta["val"]
+        quant_val = quant_node.meta["val"]
+        fused_node.meta["val"] = (None, quant_val[1], rms_val[1], rms_val[2])
+
+        find_getitem(rms_node, 2).replace_all_uses_with(residual_node_new)
+        find_getitem(quant_node, 1).replace_all_uses_with(result_node_new)
+
+    # Finally, remove matched nodes
+    graph.eliminate_dead_code()
```

Comment:
Have you verified this gets rid of all the old crud code?

Reply:
Yep, that's what the assert is for, if you're referring to the pattern we replaced.